xmppserver/src/main/java/org/jivesoftware/openfire/ldap/LdapManager.java

```diff
@@ -2142,6 +2247,144 @@ public void setIsEnclosingDNs(boolean enable) {
         return results;
     }
 
+    /**
+     * Generic routine for retrieving a single element from the LDAP server.  It's meant to be very
+     * flexible so that just about any query for a single results can make use of it without having
+     * to reimplement their own calls to LDAP.
+     * <p>
+     * The passed in filter string needs to be pre-prepared! In other words, nothing will be changed
+     * in the string before it is used as a string.
+     *
+     * @param attribute             LDAP attribute to be pulled from each result and placed in the return results.
+     *                              Typically pulled from this manager. Null means the the absolute DN is returned.
+     * @param searchFilter          Filter to use to perform the search.  Typically pulled from this manager.
+     * @param failOnMultipleResults It true, an {@link IllegalStateException} will be thrown, if the
+     *                              search result is not unique. If false, just the first result will be returned.
+     * @return A single string.
+     */
+    public String retrieveSingle(String attribute, String searchFilter, boolean failOnMultipleResults) {
```

Comment:
I /think/ this is returning a single attribute of a search result? If so, do you think you could rename it (and it's overloaded method) retrieveSingleAttribute.

Also, what's the (practical) use case for failOnMultipleResults = false? It looks to me like it's only ever set to true, not sure I can see a reason why false might be useful.

Reply:
(also, should these methods be private? not sure they need to be exposed outside this class)

Reply:
I used this name because the other function for multiple values of an attribute is called `retrieveList` and not `retrieveListAttribute`. I just want to be consistent. Should I still rename it?

Yes, `failOnMultipleResults` is always true until now. But this function is a "Generic routine for retrieving a single element from the LDAP server.". So maybe later someone may be interested in just one value of the attribute, even if there are multiple values.

It is `public` because it is a tool function like `retrieveList`. At minimum it could be package protected as it is used by the `LdapGroupProvider`.